File: QuickSort/quickSort.cpp
#include <stdio.h>
#include <conio.h>

int divide(int array[], int left, int right)
{
    int pivot = array[left];
    int l = left, r = right;
    while (l < r)
    {
        while (array[l] >= pivot)
        {
            ++l;
        }
        while (array[r] <= pivot)
        {
            --r;
        }
        if (l < r)
        {
            array[l] = array[l] + array[r];
            array[r] = array[l] - array[r];
            array[l] = array[l] - array[r];
        }
        array[left] = array[left] + array[r];
        array[r] = array[left] - array[r];
        array[left] = array[left] - array[r];
        return r;
    }
}

void QuickSort(int array[], int left, int right)
{
    if (left < right)
    {
        int p = divide(array, left, right);
        QuickSort(array, left, p);
        QuickSort(array, p + 1, right);
    }
}

int main()
{
    int i = 0, len;
    printf("\nEnter Length Of Data: ");
    scanf("%d", &len);
    int array[len];
    while (i < len && scanf("%d", &array[i]) == 1)
    {
        i++;
    };
    printf("\nAfter InsertionSort(): ");
    QuickSort(array, 0, 5);
    i = 0;
    while (i < 5)
    {
        printf("%d ", array[i]);
        i++;
    }
    return 0;
}

File: QuickSort/quickSort.py

# placing pivot to the right place
def partition(arr, left, right):
    i = left
    j = right - 1
    pivot = arr[right]

    while i < j:

        while i < right and arr[i] < pivot:
            i += 1

        while j >= i and arr[j] >= pivot:
            j -= 1

        if i < j:
            arr[i], arr[j] = arr[j], arr[i]

    if arr[i] > pivot:
        arr[i], arr[right] = arr[right], arr[i]

    return i


# for each Sub Problem / partition
def quickSort(arr, left, right):
    if left < right:
        part = partition(arr, left, right)
        quickSort(arr, left, part - 1)
        quickSort(arr, part + 1, right)


arr = [2, 3, 1, 5, 3, -1, 7, 4, 8]
quickSort(arr, 0, 8)
print(arr)
